zebra-network/src/peer/handshake.rs

```diff
@@ -222,6 +222,59 @@ impl ConnectedAddr {
             Isolated => "Isol",
         }
     }
+
+    /// Returns a list of alternate remote peer addresses, which can be used for
+    /// reconnection attempts.
+    ///
+    /// Uses the connected address, and the remote canonical address.
+    ///
+    /// Skips duplicates. If this is an outbound connection, also skips the
+    /// remote address that we're currently connected to.
+    pub fn get_alternate_addrs(
+        &self,
+        mut canonical_remote: SocketAddr,
+    ) -> impl Iterator<Item = SocketAddr> {
+        let addrs = match self {
+            OutboundDirect { addr } => {
+                // Fixup unspecified addresses and ports using known good data
+                if canonical_remote.ip().is_unspecified() {
+                    canonical_remote.set_ip(addr.ip());
+                }
+                if canonical_remote.port() == 0 {
+                    canonical_remote.set_port(addr.port());
+                }
+
+                // Try the canonical remote address, if it is different from the
+                // outbound address (which we already have in our address book)
+                if &canonical_remote != addr {
+                    vec![canonical_remote]
+                } else {
+                    Vec::new()
+                }
```

Comment:
I don't quite grok this, if the canonical_remote and self are the same, shouldn't we return vec![addr]?

Reply:
We already have the outbound address in our address book, so we didn't learn a new address from the handshake.

Does this code comment help?
https://github.com/ZcashFoundation/zebra/pull/2123/files#r633936410

(And after #2160 it would just be ignored if we tried to add it anyway.)

Reply:
Yes it does, thanks!